### PR TITLE
TEE Latency benchmark

### PR DESCRIPTION
**Pull request by igoropaniuk**
Implementation of latency benchmark within different OP-TEE layers

**Quick description:**
1. Adding timestamps is performed within optee_client, optee linux kmod, optee_os, libutee
It's also possible to add timestamps within Rich NW app and SW TA.
2. ARM PMU cycle counter is used as a common reference between both SW/NW. Value is received with a minimal latency (it's directly read from PMU registers, no API is used).
3. The fifth parameter is used to by-pass timestamp buffer within all OP-TEE layers.
4. Works on both aarch32 and aarch64 (tested on QEMU, QEMUv8 setups)
5. Variance calculation is performed for receiving more precise latency values 
6. Added printing of latency statistics for an each TEEC_InvokeCommand call. You can enable it in common.mk
`CFG_TEE_PRINT_LATENCY_STAT     ?= y`

**How to test:**
1. Enable latency benchmarks in build/common.mk
```
CFG_TEE_BENCHMARK             ?= y
CFG_TEE_PRINT_LATENCY_STAT    ?= y
 ```
2. Run latency benchmark:
`$ ./xtest -t benchmark 3001`

All test cases in xtest successfully passes on both QEMU/QEMUv8.

Relates to: 
https://github.com/linaro-swg/linux/pull/27
https://github.com/OP-TEE/build/pull/81
https://github.com/OP-TEE/optee_test/pull/145
https://github.com/OP-TEE/optee_os/pull/1175
https://github.com/OP-TEE/optee_client/pull/63

`Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

### CODE REVIEW

File: ta/latency_perf/ta_entry.c [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+/* Called each time an instance is destroyed */
+void TA_DestroyEntryPoint(void)
+{
+}
+
+/* Called each time a session is opened */
+TEE_Result TA_OpenSessionEntryPoint(uint32_t nParamTypes,
+				    TEE_Param pParams[4],
```
Comment (jenswi-linaro) [inline ta/latency_perf/ta_entry.c]: `TEE_NUM_PARAMS`?

File: host/xtest/xtest_benchmark_3000.c [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+	printf("Results:\n");
+	printf("Min=%" PRIu64 " cycles; Max=%" PRIu64
+			" cycles; Med=%" PRIu64 " cycles;\n",
+	       ((uint64_t)stats.min) * BENCH_DIVIDER,
+		   ((uint64_t)stats.max) * BENCH_DIVIDER,
+		   ((uint64_t)stats.m) * BENCH_DIVIDER);
+
+	close_latency_ta();
+
```
Comment (etienne-lms) [inline host/xtest/xtest_benchmark_3000.c]: minor: empty line

File: host/xtest/xtest_benchmark_3000.c [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		update_stats(&stats, ccnt_diff);
+	}
+
+	printf("Results:\n");
+	printf("Min=%" PRIu64 " cycles; Max=%" PRIu64
+			" cycles; Med=%" PRIu64 " cycles;\n",
+	       ((uint64_t)stats.min) * BENCH_DIVIDER,
+		   ((uint64_t)stats.max) * BENCH_DIVIDER,
+		   ((uint64_t)stats.m) * BENCH_DIVIDER);
```
Comment (etienne-lms) [inline host/xtest/xtest_benchmark_3000.c]: minor: check indent

File: host/xtest/xtest_benchmark_3000.c [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_NONE,
+					 TEEC_NONE,
+					 TEEC_NONE, TEEC_NONE);
+
+	op.params[TEE_BENCH_DEF_PARAM].memref.parent =
+				&ringbuf_shm;
+	op.params[TEE_BENCH_DEF_PARAM].memref.offset = 0;
+	op.params[TEE_BENCH_DEF_PARAM].memref.size =
+				TEE_BENCH_RB_SIZE;
```
Comment (etienne-lms) [inline host/xtest/xtest_benchmark_3000.c]: minor: useless line breaks

File: host/xtest/xtest_benchmark_3000.c [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+{
+	TEEC_ReleaseSharedMemory(&ringbuf_shm);
+	TEEC_CloseSession(&sess);
+	TEEC_FinalizeContext(&ctx);
+}
+
+static uint64_t get_timestamp(void *ringbuffer, uint64_t src)
+{
+	struct tee_ringbuf *ringb = (struct tee_ringbuf *) ringbuffer;
+	for (uint64_t ts_i = 0; ts_i < ringb->tm_ind; ts_i++) {
```
Comment (etienne-lms) [inline host/xtest/xtest_benchmark_3000.c]: minor: add empty line  + spurious space in cast.

File: host/xtest/xtest_benchmark_3000.c [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+/*
+ * TEE client stuff
+ */
+
+static TEEC_Context ctx;
+static TEEC_Session sess;
+
+static TEEC_SharedMemory ringbuf_shm = {
+		.flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT,
+		.buffer = NULL,
```
Comment (etienne-lms) [inline host/xtest/xtest_benchmark_3000.c]: minor: no need to set `buffer = NULL,`

File: host/xtest/Makefile [hunk 7]
```diff
@@ -50,6 +50,12 @@ srcs := xtest_1000.c \
 	adbg/src/security_utils_hex.c \
 	adbg/src/security_utils_mem.c
 
+
```
Comment (etienne-lms) [inline host/xtest/Makefile]: minor: empty line

File: Makefile [hunk 8]
```diff
@@ -30,13 +30,15 @@ xtest:
 			     --no-builtin-variables \
 			     q=$(q) \
 			     O=$(out-dir)/xtest \
+                 CFG_TEE_BENCHMARK=$(CFG_TEE_BENCHMARK) \
```
Comment (etienne-lms) [inline Makefile]: minor: check indentation

File: host/xtest/xtest_benchmark_3000.c [hunk 9]
```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (c) 2015, Linaro Limited
```
Comment (lackan) [inline host/xtest/xtest_benchmark_3000.c]: s/2015/2016, and also some are the same below.

File: host/xtest/xtest_benchmark_3000.c [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+
+static void print_latency_stats(void *timebuffer, struct statistics *stats)
+{
+	struct tee_time_buf *timeb = (struct tee_time_buf *)timebuffer;
+	uint64_t start = 0;
+
+	printf("Latency statistics:\n");
+	printf("===============================================================");
+	printf("===============================================================\n");
+	for (uint32_t ts_i = 0; ts_i < timeb->tm_ind; ts_i++) {
```
Comment (lackan) [inline host/xtest/xtest_benchmark_3000.c]: ts_i should be uint64_t to keep consistent with timeb->tm_ind?

File: host/xtest/xtest_benchmark_3000.c [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+		if (!ts_i)
+			start = timeb->stamps[ts_i].cnt;
+
+		printf("| CCNT=%14" PRIu64 " | SRC=%-8s | PC=0x%016"
+			PRIx64 " | Min=%14" PRIu64 " | Max=%14" PRIu64
+			" | Med=%14" PRIu64 " |;\n",
+			(timeb->stamps[ts_i].cnt-start) * BENCH_DIVIDER,
```
Comment (lackan) [inline host/xtest/xtest_benchmark_3000.c]: minor: space between '-'

File: host/xtest/xtest_benchmark_3000.c [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+static void xtest_tee_benchmark_3001(ADBG_Case_t *Case_p);
+static void xtest_tee_benchmark_3002(ADBG_Case_t *Case_p);
+
+static void xtest_tee_benchmark_3001(ADBG_Case_t *c)
+{
+	TEEC_Operation op;
+	TEEC_Result res;
+	uint32_t err_origin;
+	uint64_t ccnt_diff;
+	struct statistics stats;
```
Comment (lackan) [inline host/xtest/xtest_benchmark_3000.c]: stats have been used before initialized, may get an unexpected result.

File: host/xtest/xtest_benchmark_3000.c [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+
+}
+
+static void xtest_tee_benchmark_3002(ADBG_Case_t *c)
+{
+	TEEC_Operation op;
+	TEEC_Result res;
+	uint32_t err_origin;
+	int64_t ccnt_diff;
+	struct statistics stats[TEE_BENCH_MAX_STAMPS-1];
```
Comment (lackan) [inline host/xtest/xtest_benchmark_3000.c]: stats haven't been initialized.

File: host/xtest/xtest_benchmark_3000.c [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+	open_latency_ta();
+	memset(&op, 0, sizeof(op));
+	memset(stats, 0, sizeof(stats));
+
+	res = TEEC_AllocateSharedMemory(&ctx, &timebuf_shm);
+	tee_check_res(res, "TEEC_AllocateSharedMemory");
+
+	timeb = (struct tee_time_buf *)timebuf_shm.buffer;
+
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_NONE,
```
Comment (lackan) [inline host/xtest/xtest_benchmark_3000.c]: minor： why two TEEC_NONEs a line?

File: host/xtest/xtest_benchmark_3000.c (hunk 13 above)
Comment (igoropaniuk) [inline host/xtest/xtest_benchmark_3000.c]: see line 191
`memset(stats, 0, sizeof(stats));`

File: host/xtest/xtest_helpers.c [hunk 15]
```diff
@@ -408,3 +409,44 @@ int ree_fs_get_ta_dirname(TEEC_UUID *p_uuid, char *buffer, uint32_t len)
 			p_uuid->clockSeqAndNode[6],
 			p_uuid->clockSeqAndNode[7]);
 }
+
+/* Misc auxilary functions */
+void tee_errx(const char *msg, TEEC_Result res)
+{
+	fprintf(stderr, "%s: 0x%08x", msg, res);
```
Comment (igoropaniuk) [inline host/xtest/xtest_helpers.c]: Unfortunately, I didn't get what exactly you meant. Could you rephrase?

**Comment by igoropaniuk** [comment]
I think I've addressed all comments so far and now think this is ready for a final review and if there are no more comments we should consider merging these PRs, listed in the description.

File: ta/latency_perf/include/ta_latency_perf.h [hunk 16]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2015, Linaro Limited
```
Comment (lackan) [inline ta/latency_perf/include/ta_latency_perf.h]: minor :s/2015/2016, and the same below.

File: ta/os_test/os_test.c [hunk 17]
```diff
@@ -406,7 +406,7 @@ static TEE_Result test_mem_access_right(uint32_t param_types,
 	TEE_Result res;
 	uint32_t ret_orig;
 	uint32_t l_pts;
-	TEE_Param l_params[4] = { { {0} } };
+	TEE_Param l_params[TEE_NUM_PARAMS] = { { {0} } };
```
Comment (lackan) [inline ta/os_test/os_test.c]: Why you change 4 to TEE_NUM_PARAMS here? Does it relate to this patch?

Comment (igoropaniuk) [inline ta/os_test/os_test.c]: Because it's used in TEE_InvokeTACommand
in 
```
res = TEE_InvokeTACommand(sess, 0, TA_OS_TEST_CMD_PARAMS_ACCESS,
				  l_pts, l_params, &ret_orig);
```
And implementation of "RPC" (if I can call it sich in such a way) to another TA uses the same code for handling params in teecore/libutee, as for generic SMC call from NW. 
Without this change xtest os_test will fail.

Comment (lackan) [inline ta/os_test/os_test.c]: OK

File: host/xtest/xtest_helpers.c (hunk 15 above)
Comment (etienne-lms) [inline host/xtest/xtest_helpers.c]: Forget this about 'error origin'. does not apply to all cases where this routine can be called.

Minor change: add `\n` at string end. it was missing in the original routine.
